install-swoole.sh

```diff
@@ -1,52 +1,112 @@
 
-    if [ ! -f "${PLATFORM_CACHE_DIR}/swoole/modules/swoole.so" ]; then
-        ensure_source
-        checkout_version "$1"
-        compile_source
+    local version_prefix="v${swoole_version}-php${php_version}"
+    version_prefix="${version_prefix//\./_}"
+
+    if [ ! -f "${PLATFORM_CACHE_DIR}/${swoole_type}/swoole-src/modules/${version_prefix}-${swoole_type}.so" ]; then
+        ensure_source "$swoole_type"
+        checkout_version "${PLATFORM_CACHE_DIR}/${swoole_type}/swoole-src" "v${swoole_version}"
+        compile_source "${PLATFORM_CACHE_DIR}/${swoole_type}"
+        move_extension "$swoole_type" "$version_prefix"
     fi
 
-    copy_lib
-    enable_lib
+    copy_lib "$2" "$version_prefix"
+    enable_lib "$2"
 }
 
 enable_lib() {
     # Tell PHP to enable the extension.
-    echo "Enabling Swoole extension."
-    echo "extension=${PLATFORM_APP_DIR}/swoole.so" >> $PLATFORM_APP_DIR/php.ini
+    local swoole_type="$1"
+
+    echo "---------------------------------"
+    echo "Enabling ${swoole_type} extension."
+    echo -e "\nextension=${PLATFORM_APP_DIR}/${swoole_type}.so" >> "${PLATFORM_APP_DIR}/php.ini"
 }
 
 copy_lib() {
     # Copy the compiled library to the application directory.
-    echo "Installing Swoole extension."
-    cp $PLATFORM_CACHE_DIR/swoole/modules/swoole.so $PLATFORM_APP_DIR
+    local swoole_type="$1"
+    local version_prefix="$2"
+
+    echo "---------------------------------"
+    echo "Installing ${swoole_type} extension."
+    cp "${PLATFORM_CACHE_DIR}/${swoole_type}/swoole-src/modules/${version_prefix}-${swoole_type}.so" "${PLATFORM_APP_DIR}/${swoole_type}.so"
 }
 
 checkout_version () {
+    local checkout_dir="$1"
+    local version="$2"
     # Check out the specific Git tag that we want to build.
-    git checkout "$1"
+    cd "$checkout_dir"
+    git checkout "$version"
 }
 
 ensure_source() {
     # Ensure that the extension source code is available and up to date.
-    if [ -d "swoole" ]; then
-        cd swoole || exit 1;
+    local swoole_type="$1"
+
+    mkdir -p "${PLATFORM_CACHE_DIR}/${swoole_type}"
+    cd "${PLATFORM_CACHE_DIR}/${swoole_type}"
+
+    if [ -d "swoole-src" ]; then
+        cd swoole-src || exit 1;
+        git fetch --all --prune
+    else
+        git clone "https://github.com/${swoole_type}/swoole-src.git"
+        cd swoole-src || exit 1;
+    fi
+
+    cd "${PLATFORM_CACHE_DIR}/${swoole_type}";
+
+    if [ -d "valgrind" ]; then
+        cd valgrind || exit 1
         git fetch --all --prune
     else
-        git clone https://github.com/swoole/swoole-src.git swoole
-        cd swoole || exit 1;
+        git clone git://sourceware.org/git/valgrind.git valgrind
+        cd valgrind || exit 1
     fi
 }
 
 compile_source() {
     # Compile the extension.
+    local base_dir="$1"
+
+    echo "Compiling valgrind"
+    cd "${base_dir}/valgrind"
+    ./autogen.sh
+    ./configure --prefix="${base_dir}/swoole-src"
+    make
+    make install
```

Comment:
Valgrind was not present in the build container, causing compilation of Swoole to fail. Compiling it and installing it to the open/swoole source code allows us to compile against it.